File: sdks/python/kyuubiki_sdk/model_collaboration.py
from __future__ import annotations

import json
from collections.abc import Mapping
from typing import Any

from .errors import ModelCollaborationValidationError


MODEL_COLLABORATION_SCHEMA_VERSION = "kyuubiki.model-collaboration/v1"
MODEL_WORKFLOW_PROPOSAL_SCHEMA_VERSION = "kyuubiki.model-workflow-proposal/v1"
MODEL_HEADLESS_PLAN_SCHEMA_VERSION = "kyuubiki.model-headless-plan/v1"
MODEL_PROVIDERS = ("openai", "openai_chat", "anthropic", "gemini", "canonical")


def default_model_collaboration_policy() -> dict[str, Any]:
    return {
        "allowed_actions": [],
        "allowed_categories": [],
        "max_steps": 12,
        "max_context_bytes": 64 * 1024,
        "service_only": True,
        "allow_sensitive": False,
        "allow_destructive": False,
    }


def headless_model_tools(policy: Mapping[str, Any] | None = None) -> list[dict[str, Any]]:
    normalized = _validate_policy(policy)
    return [tool for tool in _base_model_tools() if _policy_allows_tool(normalized, tool)]


def build_model_collaboration_request(
    provider: str,
    session: Mapping[str, Any],
    context: Any,
) -> dict[str, Any]:
    normalized_session = _validate_session(session)
    if provider not in MODEL_PROVIDERS:
        _fail(f"unsupported model provider: {provider}")
    sanitized, redacted_paths = sanitize_model_context(context)
    context_bytes = len(
        json.dumps(sanitized, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    )
    policy = normalized_session["policy"]
    if context_bytes > policy["max_context_bytes"]:
        _fail(
            f"sanitized model context uses {context_bytes} bytes; policy allows "
            f"{policy['max_context_bytes']}"
        )
    tools = headless_model_tools(policy)
    if not tools:
        _fail("model collaboration policy exposes no Headless tools")
    return {
        "schema_version": MODEL_COLLABORATION_SCHEMA_VERSION,
        "provider": provider,
        "session": normalized_session,
        "instructions": [
            f"Plan only for this objective: {normalized_session['objective']}",
            "Use only supplied Headless tools and never invent action names.",
            f"Return no more than {policy['max_steps']} tool calls.",
            "Return tool calls as an untrusted proposal; never claim that execution occurred.",
        ],
        "context": sanitized,
        "redacted_paths": redacted_paths,
        "tools": project_model_tools(provider, tools),
        "output_contract": MODEL_WORKFLOW_PROPOSAL_SCHEMA_VERSION,
    }


def build_model_headless_plan(
    session: Mapping[str, Any], proposal: Mapping[str, Any]
) -> dict[str, Any]:
    normalized_session = _validate_session(session)
    policy = normalized_session["policy"]
    available = {tool["action"]: tool for tool in headless_model_tools(policy)}
    issues: list[str] = []
    if not isinstance(proposal, Mapping):
        _fail("model proposal must be a JSON object")
    if proposal.get("schema_version") != MODEL_WORKFLOW_PROPOSAL_SCHEMA_VERSION:
        issues.append(f"unsupported proposal schema_version: {proposal.get('schema_version')}")
    if proposal.get("session_id") != normalized_session["session_id"]:
        issues.append("proposal session_id does not match collaboration session")
    calls = proposal.get("calls", [])
    if not isinstance(calls, list):
        calls = []
        issues.append("model proposal calls must be an array")
    if not calls:
        issues.append("model proposal contains no tool calls")
    if len(calls) > policy["max_steps"]:
        issues.append(
            f"model proposal contains {len(calls)} calls; policy allows {policy['max_steps']}"
        )
    steps = [
        _plan_step(index, call, available, issues)
        for index, call in enumerate(calls, start=1)
    ]
    issues = sorted(set(issues))
    return {
        "schema_version": MODEL_HEADLESS_PLAN_SCHEMA_VERSION,
        "session_id": normalized_session["session_id"],
        "workflow_id": normalized_session["workflow_id"],
        "ok": not issues,
        "ready_without_confirmation": not issues
        and all(not step["requires_confirmation"] for step in steps),
        "issues": issues,
        "steps": steps,
    }


def project_model_tools(provider: str, tools: list[dict[str, Any]]) -> list[dict[str, Any]]:
    if provider not in MODEL_PROVIDERS:
        _fail(f"unsupported model provider: {provider}")
    definitions = []
    for tool in tools:
        parameters = _tool_parameters(tool["required_payload_keys"])
        if provider == "openai":
            definition = {
                "type": "function",
                "name": tool["action"],
                "description": tool["description"],
                "parameters": parameters,
                "strict": False,
            }
        elif provider == "openai_chat":
            definition = {
                "type": "function",
                "function": {
                    "name": tool["action"],
                    "description": tool["description"],
                    "parameters": parameters,
                    "strict": False,
                },
            }
        elif provider == "anthropic":
            definition = {
                "name": tool["action"],
                "description": tool["description"],
                "input_schema": parameters,
            }
        elif provider == "gemini":
            definition = {
                "name": tool["action"],
                "description": tool["description"],
                "parameters": parameters,
            }
        else:
            definition = dict(tool)
        definitions.append(definition)
    return [{"functionDeclarations": definitions}] if provider == "gemini" else definitions


def normalize_model_response(
    provider: str, session_id: str, response: Mapping[str, Any]
) -> dict[str, Any]:
    if provider == "canonical":
        return _normalize_canonical(session_id, response)
    calls: list[dict[str, Any]] = []
    summaries: list[str] = []
    if provider in ("openai", "openai_chat"):
        _collect_openai(response, calls, summaries)
    elif provider == "anthropic":
        _collect_anthropic(response, calls, summaries)
    elif provider == "gemini":
        _collect_gemini(response, calls, summaries)
    else:
        _fail(f"unsupported model provider: {provider}")
    if not calls:
        _fail("provider response contains no supported tool calls")
    return {
        "schema_version": MODEL_WORKFLOW_PROPOSAL_SCHEMA_VERSION,
        "session_id": session_id,
        "summary": "\n".join(summaries),
        "calls": calls,
    }


def sanitize_model_context(context: Any) -> tuple[Any, list[str]]:
    redacted_paths: list[str] = []

    def sanitize(value: Any, path: str) -> Any:
        if isinstance(value, Mapping):
            result = {}
            for key, child in value.items():
                text_key = str(key)
                next_path = f"{path}/{_escape_pointer(text_key)}"
                if _sensitive_key(text_key):
                    redacted_paths.append(next_path)
                    result[text_key] = "[REDACTED]"
                else:
                    result[text_key] = sanitize(child, next_path)
            return result
        if isinstance(value, list):
            return [sanitize(child, f"{path}/{index}") for index, child in enumerate(value)]
        if isinstance(value, str) and value.lstrip().lower().startswith("bearer "):
            redacted_paths.append(path or "/")
            return "[REDACTED]"
        return value

    return sanitize(context, ""), redacted_paths


def _normalize_policy(policy: Mapping[str, Any] | None) -> dict[str, Any]:
    normalized = default_model_collaboration_policy()
    if policy is not None:
        if not isinstance(policy, Mapping):
            _fail("model collaboration policy must be a JSON object")
        normalized.update(policy)
    return normalized


def _validate_session(session: Mapping[str, Any]) -> dict[str, Any]:
    if not isinstance(session, Mapping):
        _fail("model collaboration session must be a JSON object")
    errors = []
    if session.get("schema_version") != MODEL_COLLABORATION_SCHEMA_VERSION:
        errors.append(f"unsupported session schema_version: {session.get('schema_version')}")
    for key in ("session_id", "workflow_id", "objective", "created_at"):
        value = session.get(key)
        if not isinstance(value, str) or not value.strip():
            errors.append(f"{key} is required")
    policy = _validate_policy(session.get("policy"))
    for key in ("max_steps", "max_context_bytes"):
        value = policy.get(key)
        if not isinstance(value, int) or isinstance(value, bool) or value <= 0:
            errors.append(f"{key} must be a positive integer")
    for key in ("allowed_actions", "allowed_categories"):
        if not isinstance(policy.get(key), list):
            errors.append(f"{key} must be an array")
    if errors:
        raise ModelCollaborationValidationError(errors)
    normalized = dict(session)
    normalized["language"] = session.get("language", "en")
    normalized["policy"] = policy
    return normalized


def _validate_policy(policy: Mapping[str, Any] | None) -> dict[str, Any]:
    normalized = _normalize_policy(policy)
    errors = []
    for key in ("allowed_actions", "allowed_categories"):
        value = normalized.get(key)
        if not isinstance(value, list) or not all(isinstance(item, str) for item in value):
            errors.append(f"{key} must be an array of strings")
    for key in ("service_only", "allow_sensitive", "allow_destructive"):
        if not isinstance(normalized.get(key), bool):
            errors.append(f"{key} must be a boolean")
    for key in ("max_steps", "max_context_bytes"):
        value = normalized.get(key)
        if not isinstance(value, int) or isinstance(value, bool) or value <= 0:
            errors.append(f"{key} must be a positive integer")
    if errors:
        raise ModelCollaborationValidationError(errors)
    return normalized


def _plan_step(
    index: int,
    call: Any,
    available: dict[str, dict[str, Any]],
    issues: list[str],
) -> dict[str, Any]:
    if not isinstance(call, Mapping):
        issues.append(f"step {index} must be a JSON object")
        call = {}
    action = call.get("action")
    if not isinstance(action, str) or not action:
        action = ""
        issues.append(f"step {index} action is required")
    tool = available.get(action)
    if tool is None:
        issues.append(f"step {index} action {action} is unknown or blocked by policy")
    payload = call.get("payload", {})
    if not isinstance(payload, Mapping):
        issues.append(f"step {index} ({action}) payload must be a JSON object")
        payload = {}
    if tool:
        for key in tool["required_payload_keys"]:
            if not _has_present_value(payload, key):
                issues.append(f"step {index} ({action}) is missing required payload key {key}")
        _validate_known_payload(index, action, payload, issues)
    risk = tool["risk"] if tool else "normal"
    return {
        "index": index,
        "action": action,
        "category": tool["category"] if tool else None,
        "risk": risk,
        "payload": dict(payload),
        "requires_confirmation": risk != "normal",
        "confirmation_reason": _confirmation_reason(risk),
        "output_keys": list(tool["output_keys"]) if tool else [],
    }


def _collect_openai(
    response: Mapping[str, Any], calls: list[dict[str, Any]], summaries: list[str]
) -> None:
    for item in _map_items(response.get("output")):
        if item.get("type") == "function_call":
            calls.append(_parse_call(item.get("call_id") or item.get("id"), item.get("name"), item.get("arguments")))
        elif item.get("type") == "message":
            for part in _map_items(item.get("content")):
                if part.get("type") in ("output_text", "text"):
                    _collect_text(part.get("text"), summaries)
    for choice in _map_items(response.get("choices")):
        message = choice.get("message", {}) if isinstance(choice.get("message"), Mapping) else {}
        _collect_text(message.get("content"), summaries)
        for item in _map_items(message.get("tool_calls")):
            function = item.get("function", {}) if isinstance(item.get("function"), Mapping) else {}
            calls.append(_parse_call(item.get("id"), function.get("name"), function.get("arguments")))


def _collect_anthropic(
    response: Mapping[str, Any], calls: list[dict[str, Any]], summaries: list[str]
) -> None:
    for block in _map_items(response.get("content")):
        if block.get("type") == "tool_use":
            calls.append(_parse_call(block.get("id"), block.get("name"), block.get("input")))
        elif block.get("type") == "text":
            _collect_text(block.get("text"), summaries)


def _collect_gemini(
    response: Mapping[str, Any], calls: list[dict[str, Any]], summaries: list[str]
) -> None:
    for candidate in _map_items(response.get("candidates")):
        content = candidate.get("content", {}) if isinstance(candidate.get("content"), Mapping) else {}
        for part in _map_items(content.get("parts")):
            _collect_text(part.get("text"), summaries)
            function = part.get("functionCall")
            if function:
                calls.append(_parse_call(function.get("id"), function.get("name"), function.get("args")))
    for step in _map_items(response.get("steps")):
        if step.get("type") == "function_call":
            calls.append(_parse_call(step.get("id"), step.get("name"), step.get("arguments")))


def _normalize_canonical(session_id: str, response: Mapping[str, Any]) -> dict[str, Any]:
    if response.get("session_id") != session_id:
        _fail("canonical proposal session_id does not match requested session")
    if response.get("schema_version") != MODEL_WORKFLOW_PROPOSAL_SCHEMA_VERSION:
        _fail("canonical proposal has unsupported schema_version")
    calls = []
    for call in response.get("calls", []):
        if not isinstance(call, Mapping):
            _fail("canonical proposal calls must be JSON objects")
        normalized = dict(call)
        normalized["payload"] = _parse_arguments(call.get("payload", {}))
        calls.append(normalized)
    proposal = dict(response)
    proposal["calls"] = calls
    return proposal


def _parse_call(call_id: Any, name: Any, arguments: Any) -> dict[str, Any]:
    if not isinstance(name, str) or not name:
        _fail("provider tool call is missing a string name")
    if arguments is None:
        _fail(f"provider tool call {name} is missing arguments")
    return {"id": call_id if isinstance(call_id, str) else None, "action": name, "payload": _parse_arguments(arguments), "reason": None}


def _parse_arguments(arguments: Any) -> dict[str, Any]:
    if isinstance(arguments, str):
        try:
            arguments = json.loads(arguments)
        except json.JSONDecodeError as error:
            _fail(f"provider tool arguments are invalid JSON: {error.msg}")
    if not isinstance(arguments, Mapping):
        _fail("provider tool arguments must decode to a JSON object")
    return dict(arguments)


def _tool_parameters(required_keys: list[str]) -> dict[str, Any]:
    return {
        "type": "object",
        "properties": {key: {"description": f"Required `{key}` payload"} for key in required_keys},
        "required": required_keys,
        "additionalProperties": True,
    }


def _policy_allows_tool(policy: Mapping[str, Any], tool: Mapping[str, Any]) -> bool:
    return (
        (not policy["allowed_actions"] or tool["action"] in policy["allowed_actions"])
        and (not policy["allowed_categories"] or tool["category"] in policy["allowed_categories"])
        and (policy["allow_sensitive"] or tool["risk"] != "sensitive")
        and (policy["allow_destructive"] or tool["risk"] != "destructive")
        and (not policy["service_only"] or tool["runtime"] == "service")
    )


def _has_present_value(payload: Mapping[str, Any], key: str) -> bool:
    value = payload.get(key)
    return value is not None and (not isinstance(value, str) or bool(value.strip()))


def _validate_known_payload(
    index: int,
    action: str,
    payload: Mapping[str, Any],
    issues: list[str],
) -> None:
    string_keys: tuple[str, ...] = ()
    object_keys: tuple[str, ...] = ()
    if action in ("fem_submit", "direct_solver_rpc"):
        string_keys, object_keys = ("solve_kind",), ("payload",)
    elif action == "workflow_submit_catalog":
        string_keys, object_keys = ("workflow_id",), ("input_artifacts",)
    elif action == "workflow_submit_graph":
        object_keys = ("graph", "input_artifacts")
    elif action in ("operator_task_prepare", "operator_task_execute"):
        object_keys = ("task",)
    elif action in ("operator_task_batch_prepare", "operator_task_batch_execute"):
        object_keys = ("batch",)
    elif action in ("job_wait", "result_fetch", "job_cancel"):
        string_keys = ("job_id",)
    elif action == "result_chunk_fetch":
        string_keys = ("job_id", "kind")

    for key in string_keys:
        value = payload.get(key)
        if value is not None and (not isinstance(value, str) or not value.strip()):
            issues.append(
                f"step {index} ({action}) payload key {key} must be a non-empty string"
            )
    for key in object_keys:
        value = payload.get(key)
        if value is not None and not isinstance(value, Mapping):
            issues.append(f"step {index} ({action}) payload key {key} must be a JSON object")
    if action == "result_chunk_fetch":
        for key in ("offset", "limit"):
            value = payload.get(key)
            if value is not None and (
                not isinstance(value, int) or isinstance(value, bool) or value < 0
            ):
                issues.append(
                    f"step {index} ({action}) payload key {key} must be an unsigned integer"
                )


def _confirmation_reason(risk: str) -> str | None:
    if risk == "sensitive":
        return "sensitive Headless action requires explicit approval before dispatch"
    if risk == "destructive":
        return "destructive Headless action requires explicit approval before dispatch"
    return None


def _collect_text(value: Any, summaries: list[str]) -> None:
    if isinstance(value, str) and value.strip():
        summaries.append(value.strip())


def _map_items(value: Any) -> list[Mapping[str, Any]]:
    if not isinstance(value, list):
        return []
    return [item for item in value if isinstance(item, Mapping)]


def _sensitive_key(key: str) -> bool:
    normalized = key.lower().replace("-", "_").replace(".", "_")
    return any(marker in normalized for marker in ("token", "secret", "password", "api_key", "apikey", "authorization", "credential", "private_key"))


def _escape_pointer(value: str) -> str:
    return value.replace("~", "~0").replace("/", "~1")


def _tool(
    action: str,
    category: str,
    description: str,
    risk: str,
    runtime: str,
    required: tuple[str, ...] = (),
    outputs: tuple[str, ...] = (),
) -> dict[str, Any]:
    return {
        "action": action,
        "category": category,
        "description": description,
        "risk": risk,
        "runtime": runtime,
        "required_payload_keys": list(required),
        "output_keys": list(outputs),
    }


def _base_model_tools() -> list[dict[str, Any]]:
    service = "service"
    return [
        _tool("service_health", "discovery", "Check control-plane health.", "normal", service, outputs=("health",)),
        _tool("protocol_describe", "discovery", "Read protocol compatibility and service endpoints.", "normal", service, outputs=("protocol",)),
        _tool("agents_describe", "discovery", "List reachable agents and capabilities.", "normal", service, outputs=("agents",)),
        _tool("workflow_catalog_list", "discovery", "List centrally owned workflow templates.", "normal", service, outputs=("workflows",)),
        _tool("operator_catalog_list", "discovery", "List workflow operator descriptors.", "normal", service, outputs=("operators",)),
        _tool("fem_submit", "solve", "Submit a FEM solve kind and model payload.", "sensitive", service, ("solve_kind", "payload"), ("job",)),
        _tool("direct_solver_rpc", "solve", "Call a configured solver agent without Orchestra.", "sensitive", "direct", ("solve_kind", "payload"), ("result",)),
        _tool("workflow_submit_catalog", "workflow", "Submit a catalog workflow job.", "sensitive", service, ("workflow_id", "input_artifacts"), ("job",)),
        _tool("workflow_submit_graph", "workflow", "Submit a validated inline workflow graph.", "sensitive", service, ("graph", "input_artifacts"), ("job",)),
        _tool("operator_task_prepare", "task_ir", "Preflight one language-neutral Operator TaskIR envelope.", "normal", service, ("task",), ("preparation",)),
        _tool("operator_task_execute", "task_ir", "Execute one prepared Operator TaskIR envelope.", "sensitive", service, ("task",), ("execution",)),
        _tool("operator_task_batch_prepare", "task_ir", "Preflight an Operator TaskIR batch.", "normal", service, ("batch",), ("preparation",)),
        _tool("operator_task_batch_execute", "task_ir", "Execute an Operator TaskIR batch.", "sensitive", service, ("batch",), ("execution",)),
        _tool("job_wait", "observation", "Poll a job until it reaches a terminal state.", "normal", service, ("job_id",), ("job",)),
        _tool("result_fetch", "observation", "Fetch the retained result bundle for a job.", "normal", service, ("job_id",), ("result",)),
        _tool("result_chunk_fetch", "observation", "Fetch one bounded result chunk.", "normal", service, ("job_id", "kind"), ("chunk",)),
        _tool("job_cancel", "lifecycle", "Cancel a running job after explicit approval.", "destructive", service, ("job_id",), ("job",)),
    ]


def _fail(message: str) -> None:
    raise ModelCollaborationValidationError([message])

File: sdks/elixir/lib/kyuubiki_sdk/model_collaboration.ex
defmodule KyuubikiSdk.ModelCollaboration do
  @moduledoc "Provider-neutral model proposals over constrained Headless actions."

  alias KyuubikiSdk.Error
  alias KyuubikiSdk.ModelProviderAdapters

  @session_schema "kyuubiki.model-collaboration/v1"
  @proposal_schema "kyuubiki.model-workflow-proposal/v1"
  @plan_schema "kyuubiki.model-headless-plan/v1"

  def session_schema_version, do: @session_schema
  def proposal_schema_version, do: @proposal_schema
  def plan_schema_version, do: @plan_schema

  def default_policy do
    %{
      "allowed_actions" => [],
      "allowed_categories" => [],
      "max_steps" => 12,
      "max_context_bytes" => 64 * 1024,
      "service_only" => true,
      "allow_sensitive" => false,
      "allow_destructive" => false
    }
  end

  def tools(policy \\ %{}) when is_map(policy) do
    policy = Map.merge(default_policy(), policy)
    Enum.filter(base_tools(), &policy_allows_tool?(policy, &1))
  end

  def build_request(provider, session, context) do
    with {:ok, provider} <- ModelProviderAdapters.normalize_provider(provider),
         {:ok, session} <- validate_session(session) do
      {context, redacted_paths} = ModelProviderAdapters.sanitize_context(context)
      context_bytes = context |> Jason.encode!() |> byte_size()
      policy = session["policy"]
      available_tools = tools(policy)

      cond do
        context_bytes > policy["max_context_bytes"] ->
          validation_error(
            "sanitized model context uses #{context_bytes} bytes; policy allows #{policy["max_context_bytes"]}"
          )

        available_tools == [] ->
          validation_error("model collaboration policy exposes no Headless tools")

        true ->
          {:ok,
           %{
             "schema_version" => @session_schema,
             "provider" => provider,
             "session" => session,
             "instructions" => [
               "Plan only for this objective: #{session["objective"]}",
               "Use only supplied Headless tools and never invent action names.",
               "Return no more than #{policy["max_steps"]} tool calls.",
               "Return tool calls as an untrusted proposal; never claim that execution occurred."
             ],
             "context" => context,
             "redacted_paths" => redacted_paths,
             "tools" => ModelProviderAdapters.project_tools(provider, available_tools),
             "output_contract" => @proposal_schema
           }}
      end
    end
  end

  def build_plan(session, proposal) when is_map(proposal) do
    with {:ok, session} <- validate_session(session) do
      policy = session["policy"]
      available = Map.new(tools(policy), &{&1["action"], &1})
      {calls, initial_issues} = proposal_calls(proposal)

      issues =
        initial_issues
        |> add_issue(
          proposal["schema_version"] != @proposal_schema,
          "unsupported proposal schema_version: #{inspect(proposal["schema_version"])}"
        )
        |> add_issue(
          proposal["session_id"] != session["session_id"],
          "proposal session_id does not match collaboration session"
        )
        |> add_issue(calls == [], "model proposal contains no tool calls")
        |> add_issue(
          length(calls) > policy["max_steps"],
          "model proposal contains #{length(calls)} calls; policy allows #{policy["max_steps"]}"
        )

      {steps, issues} =
        calls
        |> Enum.with_index(1)
        |> Enum.map_reduce(issues, fn {call, index}, current_issues ->
          plan_step(index, call, available, current_issues)
        end)

      issues = issues |> Enum.uniq() |> Enum.sort()

      {:ok,
       %{
         "schema_version" => @plan_schema,
         "session_id" => session["session_id"],
         "workflow_id" => session["workflow_id"],
         "ok" => issues == [],
         "ready_without_confirmation" =>
           issues == [] and Enum.all?(steps, &(not &1["requires_confirmation"])),
         "issues" => issues,
         "steps" => steps
       }}
    end
  end

  def build_plan(_session, _proposal),
    do: validation_error("model proposal must be a JSON object")

  defdelegate project_tools(provider, available_tools), to: ModelProviderAdapters
  defdelegate normalize_response(provider, session_id, response), to: ModelProviderAdapters
  defdelegate sanitize_context(context), to: ModelProviderAdapters

  defp validate_session(session) when is_map(session) do
    raw_policy = session["policy"]
    policy = normalize_policy(raw_policy)

    errors =
      []
      |> add_issue(
        session["schema_version"] != @session_schema,
        "unsupported session schema_version: #{inspect(session["schema_version"])}"
      )
      |> require_session_string(session, "session_id")
      |> require_session_string(session, "workflow_id")
      |> require_session_string(session, "objective")
      |> require_session_string(session, "created_at")
      |> require_positive_integer(policy, "max_steps")
      |> require_positive_integer(policy, "max_context_bytes")
      |> require_list(policy, "allowed_actions")
      |> require_list(policy, "allowed_categories")
      |> add_policy_issues(raw_policy)

    if errors == [] do
      {:ok,
       session
       |> Map.put_new("language", "en")
       |> Map.put("policy", policy)}
    else
      {:error, Error.model_collaboration_validation(errors)}
    end
  end

  defp validate_session(_session),
    do: validation_error("model collaboration session must be a JSON object")

  defp normalize_policy(nil), do: default_policy()

  defp normalize_policy(policy) when is_map(policy) do
    Enum.reduce(default_policy(), default_policy(), fn {key, default}, normalized ->
      value = Map.get(policy, key, default)
      Map.put(normalized, key, if(valid_policy_value?(value, default), do: value, else: default))
    end)
  end

  defp normalize_policy(_policy), do: default_policy()

  defp plan_step(index, call, available, issues) when is_map(call) do
    action = call["action"]
    tool = if is_binary(action), do: available[action], else: nil
    payload = if is_map(call["payload"]), do: call["payload"], else: %{}

    issues =
      issues
      |> add_issue(not is_binary(action) or action == "", "step #{index} action is required")
      |> add_issue(
        is_nil(tool),
        "step #{index} action #{action || ""} is unknown or blocked by policy"
      )
      |> add_issue(
        not is_map(call["payload"]),
        "step #{index} (#{action || ""}) payload must be a JSON object"
      )
      |> add_missing_payload_issues(index, action, payload, tool)
      |> add_payload_shape_issues(index, action, payload, tool)

    risk = if tool, do: tool["risk"], else: "normal"

    {%{
       "index" => index,
       "action" => action || "",
       "category" => if(tool, do: tool["category"], else: nil),
       "risk" => risk,
       "payload" => payload,
       "requires_confirmation" => risk != "normal",
       "confirmation_reason" => confirmation_reason(risk),
       "output_keys" => if(tool, do: tool["output_keys"], else: [])
     }, issues}
  end

  defp plan_step(index, _call, _available, issues) do
    plan_step(index, %{}, %{}, ["step #{index} must be a JSON object" | issues])
  end

  defp proposal_calls(proposal) do
    case Map.get(proposal, "calls", []) do
      calls when is_list(calls) -> {calls, []}
      _ -> {[], ["model proposal calls must be an array"]}
    end
  end

  defp add_missing_payload_issues(issues, _index, _action, _payload, nil), do: issues

  defp add_missing_payload_issues(issues, index, action, payload, tool) do
    Enum.reduce(tool["required_payload_keys"], issues, fn key, current ->
      add_issue(
        current,
        not present_value?(payload[key]),
        "step #{index} (#{action}) is missing required payload key #{key}"
      )
    end)
  end

  defp add_payload_shape_issues(issues, _index, _action, _payload, nil), do: issues

  defp add_payload_shape_issues(issues, index, action, payload, _tool) do
    {string_keys, object_keys} =
      case action do
        action when action in ["fem_submit", "direct_solver_rpc"] ->
          {["solve_kind"], ["payload"]}

        "workflow_submit_catalog" ->
          {["workflow_id"], ["input_artifacts"]}

        "workflow_submit_graph" ->
          {[], ["graph", "input_artifacts"]}

        action when action in ["operator_task_prepare", "operator_task_execute"] ->
          {[], ["task"]}

        action when action in ["operator_task_batch_prepare", "operator_task_batch_execute"] ->
          {[], ["batch"]}

        action when action in ["job_wait", "result_fetch", "job_cancel"] ->
          {["job_id"], []}

        "result_chunk_fetch" ->
          {["job_id", "kind"], []}

        _ ->
          {[], []}
      end

    issues =
      Enum.reduce(string_keys, issues, fn key, current ->
        add_issue(
          current,
          Map.has_key?(payload, key) and
            (not is_binary(payload[key]) or String.trim(payload[key]) == ""),
          "step #{index} (#{action}) payload key #{key} must be a non-empty string"
        )
      end)

    issues =
      Enum.reduce(object_keys, issues, fn key, current ->
        add_issue(
          current,
          Map.has_key?(payload, key) and not is_map(payload[key]),
          "step #{index} (#{action}) payload key #{key} must be a JSON object"
        )
      end)

    if action == "result_chunk_fetch" do
      Enum.reduce(["offset", "limit"], issues, fn key, current ->
        value = payload[key]

        add_issue(
          current,
          not is_nil(value) and (not is_integer(value) or value < 0),
          "step #{index} (#{action}) payload key #{key} must be an unsigned integer"
        )
      end)
    else
      issues
    end
  end

  defp policy_allows_tool?(policy, tool) do
    (policy["allowed_actions"] == [] or tool["action"] in policy["allowed_actions"]) and
      (policy["allowed_categories"] == [] or tool["category"] in policy["allowed_categories"]) and
      (policy["allow_sensitive"] or tool["risk"] != "sensitive") and
      (policy["allow_destructive"] or tool["risk"] != "destructive") and
      (not policy["service_only"] or tool["runtime"] == "service")
  end

  defp require_session_string(errors, session, key) do
    add_issue(errors, not present_value?(session[key]), "#{key} is required")
  end

  defp require_positive_integer(errors, policy, key) do
    value = policy[key]
    add_issue(errors, not is_integer(value) or value <= 0, "#{key} must be a positive integer")
  end

  defp require_list(errors, policy, key) do
    add_issue(errors, not is_list(policy[key]), "#{key} must be an array")
  end

  defp add_policy_issues(errors, nil), do: errors

  defp add_policy_issues(errors, policy) when is_map(policy) do
    Enum.reduce(default_policy(), errors, fn {key, default}, current ->
      value = Map.get(policy, key, default)
      add_issue(current, not valid_policy_value?(value, default), policy_issue(key, default))
    end)
  end

  defp add_policy_issues(errors, _policy),
    do: ["model collaboration policy must be a JSON object" | errors]

  defp valid_policy_value?(value, default) when is_list(default),
    do: is_list(value) and Enum.all?(value, &is_binary/1)

  defp valid_policy_value?(value, default) when is_boolean(default), do: is_boolean(value)

  defp valid_policy_value?(value, default) when is_integer(default),
    do: is_integer(value) and value > 0

  defp policy_issue(key, default) when is_list(default), do: "#{key} must be an array of strings"
  defp policy_issue(key, default) when is_boolean(default), do: "#{key} must be a boolean"
  defp policy_issue(key, _default), do: "#{key} must be a positive integer"

  defp add_issue(errors, true, message), do: [message | errors]
  defp add_issue(errors, false, _message), do: errors

  defp present_value?(value) when is_binary(value), do: String.trim(value) != ""
  defp present_value?(value), do: not is_nil(value)

  defp confirmation_reason("sensitive"),
    do: "sensitive Headless action requires explicit approval before dispatch"

  defp confirmation_reason("destructive"),
    do: "destructive Headless action requires explicit approval before dispatch"

  defp confirmation_reason(_risk), do: nil

  defp base_tools do
    service = "service"

    [
      tool("service_health", "discovery", "Check control-plane health.", "normal", service, [], [
        "health"
      ]),
      tool(
        "protocol_describe",
        "discovery",
        "Read protocol compatibility and service endpoints.",
        "normal",
        service,
        [],
        ["protocol"]
      ),
      tool(
        "agents_describe",
        "discovery",
        "List reachable agents and capabilities.",
        "normal",
        service,
        [],
        ["agents"]
      ),
      tool(
        "workflow_catalog_list",
        "discovery",
        "List centrally owned workflow templates.",
        "normal",
        service,
        [],
        ["workflows"]
      ),
      tool(
        "operator_catalog_list",
        "discovery",
        "List workflow operator descriptors.",
        "normal",
        service,
        [],
        ["operators"]
      ),
      tool(
        "fem_submit",
        "solve",
        "Submit a FEM solve kind and model payload.",
        "sensitive",
        service,
        ["solve_kind", "payload"],
        ["job"]
      ),
      tool(
        "direct_solver_rpc",
        "solve",
        "Call a configured solver agent without Orchestra.",
        "sensitive",
        "direct",
        ["solve_kind", "payload"],
        ["result"]
      ),
      tool(
        "workflow_submit_catalog",
        "workflow",
        "Submit a catalog workflow job.",
        "sensitive",
        service,
        ["workflow_id", "input_artifacts"],
        ["job"]
      ),
      tool(
        "workflow_submit_graph",
        "workflow",
        "Submit a validated inline workflow graph.",
        "sensitive",
        service,
        ["graph", "input_artifacts"],
        ["job"]
      ),
      tool(
        "operator_task_prepare",
        "task_ir",
        "Preflight one language-neutral Operator TaskIR envelope.",
        "normal",
        service,
        ["task"],
        ["preparation"]
      ),
      tool(
        "operator_task_execute",
        "task_ir",
        "Execute one prepared Operator TaskIR envelope.",
        "sensitive",
        service,
        ["task"],
        ["execution"]
      ),
      tool(
        "operator_task_batch_prepare",
        "task_ir",
        "Preflight an Operator TaskIR batch.",
        "normal",
        service,
        ["batch"],
        ["preparation"]
      ),
      tool(
        "operator_task_batch_execute",
        "task_ir",
        "Execute an Operator TaskIR batch.",
        "sensitive",
        service,
        ["batch"],
        ["execution"]
      ),
      tool(
        "job_wait",
        "observation",
        "Poll a job until it reaches a terminal state.",
        "normal",
        service,
        ["job_id"],
        ["job"]
      ),
      tool(
        "result_fetch",
        "observation",
        "Fetch the retained result bundle for a job.",
        "normal",
        service,
        ["job_id"],
        ["result"]
      ),
      tool(
        "result_chunk_fetch",
        "observation",
        "Fetch one bounded result chunk.",
        "normal",
        service,
        ["job_id", "kind"],
        ["chunk"]
      ),
      tool(
        "job_cancel",
        "lifecycle",
        "Cancel a running job after explicit approval.",
        "destructive",
        service,
        ["job_id"],
        ["job"]
      )
    ]
  end

  defp tool(action, category, description, risk, runtime, required, outputs) do
    %{
      "action" => action,
      "category" => category,
      "description" => description,
      "risk" => risk,
      "runtime" => runtime,
      "required_payload_keys" => required,
      "output_keys" => outputs
    }
  end

  defp validation_error(message),
    do: {:error, Error.model_collaboration_validation([message])}
end
